### DISCUSSION TITLE
Partially transparent composite, with one color fully transparent.

### DISCUSSION THREAD

**Discussion by guezmo** [ImageMagick/ImageMagick#2814] [Help] [answered]
Hello, I am trying to make a partially transparent composite, with a single color fully transparent. 

It is easy enough to composite an image with transparency factor.  Also easy enough to change a color to full transparency, but how do you do both, such that a specific color remains FULLY transparent, while everything else is partially transparent?

This changes color 0600a to transparent and makes a composite:
convert A.jpg '(' B.bmp -fuzz 1% -transparent '#06000a' ')'   -composite   BoverA.jpg

This will overlay an image with 50 percent transparency:
convert A.jpg '(' B.bmp  ')' -alpha set -channel Alpha -evaluate set 50%  -composite BpverA.jpg   

How do I both BOTH such that I get the 50 percent transparency , but for color #06000a I still get 100 percent transparency?

Thanks

**Comment by fmw42** [comment]
Make a mask that is white where you want fully opaque colors and mid-gray where you want 50% transparency. Then put the mask into the alpha channel of the image.

**Comment by snibgo** [comment]
I'm unsure what the problem is.

When you composite one image (eg B.bmp, perhaps modified) over another one (eg A.jpg) then any pixels that are fully transparent in the top image will have no effect in the final result.

**Reply by guezmo** [reply]
Thanks for the reply...   Well... my command string may be wrong, but when I do the below, the second Alpha destroys the first, and the #06000a ends up only at 50 percent rather than full transparency.    I think I understand the reply from snibgo, I just have to figure out how to make the mask.   This must have to do with the alpha channel I have sometimes use (rgba), which I usually use for solid colors. 

convert A.jpg '(' B.bmp -fuzz 1% -transparent '#06000a' ')' -alpha set -channel Alpha -evaluate set 50%  -composite   BoverA.jpg

**Comment by fmw42** [comment]
Please post your input images and explain what you want for the result.

**Reply by guezmo** [reply]
Thanks for making me more aware of using a mask concept.
This accomplishes what I want, or very close to what I want, though there may be a faster way.    I guess it gets complicated if I have to do my own fuzz in the logic statements.    If using B.jpg, fuzz is a must, because of compression.   The logic part corresponds to "#06000a"

 convert A.jpg  '(' B.bmp -channel alpha -fx '( r>(6/255) || g>0 || b>(10/255) ) ? .5 : 1' ')' -alpha on  -composite  BoverA.jpg

**Reply by fmw42** [reply]
Why do you have -channel alpha in the command?  -fx is working on the r,g,b channels and not the alpha channel. Can you post your A.jpg and B.bmp so we can test and suggest other ways by thresholding the channels to make and combine masks.

**Reply by guezmo** [reply]
I think by saying -channel alpha, the fx will only change the alpha channel component?

**Comment by snibgo** [comment, marked as answer]
> ... the second Alpha destroys the first ...

You set some pixels transparent (alpha=0), and then set the alpha of all pixels (in both images) to 50%. So of course all pixels are then 50%.

Instead, you might multiply all alpha values by 0.5. Then opaque pixels (alpha=100%) will become 50%, and transparent pixels (alpha=0) remain transparent.
```
-channel A -evaluate Multiply 0.5 +channel
```

**Reply by guezmo** [reply]
Yes, this makes sense, seems to work quite well... thanks:
convert A.jpg '(' B.bmp -fuzz 1% -transparent '#06000a' ')' -channel Alpha -evaluate Multiply .5 -composite  BoverA.jpg

**Comment by guezmo** [comment]
Thanks for the help, all the replies were very helpful to think more in terms of a mask.   I included a random test image I made. 

I like this one since fuzz is easy to use and it is faster:
-bash-4.2$ time **convert A.jpg '(' B.png -fuzz 1% -transparent '#06000a' ')' -channel Alpha -evaluate Multiply .5 -composite  BoverA.jpg**  
real    0m1.779s    user    0m0.276s     sys     0m0.085s

I tried this also, though takes longer and fuzz factor harder to apply ( I just used > in this particular case):
time **convert A.jpg  '(' B.png -channel alpha -fx '( r>(6/255) || g>0 || b>(10/255) ) ? .5 : 0' ')' -alpha on  -composite  BoverA.jpg**
real    0m3.097s    user    0m18.103s     sys     0m0.087s

![A](https://user-images.githubusercontent.com/30556949/98967115-193f9580-24da-11eb-80c9-0ef6fdfacb34.jpg)
<img width="664" alt="B" src="https://user-images.githubusercontent.com/30556949/98967125-1cd31c80-24da-11eb-94e6-aa90b347b304.png">
![BoverA](https://user-images.githubusercontent.com/30556949/98967129-1f357680-24da-11eb-97e5-6c0600b5b778.jpg)

**Comment by guezmo** [comment]
Gosh, you can only mark one answer!  They all helped a lot.  I thought I could mark both of them.
